google/cloud/security/scanner/audit/org_rules_engine.py

```diff
@@ -298,35 +391,34 @@ def find_violations(self, resource, policy_binding):
         Returns:
             A generator of the rule violations.
         """
-        violations = []
+        violations = itertools.chain()
         for curr_resource in resource.get_ancestors():
-            resource_rules = self.resource_rules_map.get(curr_resource)
-
-            if not resource_rules:
-                continue
-
-            # Check whether rules match if the applies_to condition is met:
-            # SELF: check rules if the starting resource == current resource
-            # CHILDREN: check rules if starting resource != current resource
-            # SELF_AND_CHILDREN: always check rules
-            do_rules_check = (
-                (resource_rules.applies_to == RuleAppliesTo.SELF and
-                 resource == curr_resource) or
-                (resource_rules.applies_to == RuleAppliesTo.CHILDREN and
-                 resource != curr_resource) or
-                (resource_rules.applies_to == RuleAppliesTo.SELF_AND_CHILDREN))
-
-            if not do_rules_check:
-                continue
-
-            # pylint: disable=redefined-variable-type
-            violations = itertools.chain(
-                violations,
-                resource_rules.find_mismatches(resource, policy_binding))
+            resource_rules = self._get_resource_rules(curr_resource)
+
+            for resource_rule in resource_rules:
+                # Check whether rules match if the applies_to condition is met:
+                # SELF: check rules if the starting resource == current resource
+                # CHILDREN: check rules if starting resource != current resource
+                # SELF_AND_CHILDREN: always check rules
+                rule_applies_to_resource = (
+                    (resource_rule.applies_to == RuleAppliesTo.SELF and
+                     resource == curr_resource) or
+                    (resource_rule.applies_to == RuleAppliesTo.CHILDREN and
+                     resource != curr_resource) or
+                    (resource_rule.applies_to ==
+                     RuleAppliesTo.SELF_AND_CHILDREN))
+
+                if not rule_applies_to_resource:
+                    continue
+
+                violations = itertools.chain(
+                    violations,
+                    resource_rule.find_mismatches(resource, policy_binding))
+
+                # If the rule does not inherit the parents' rules, stop.
+                if not resource_rule.inherit_from_parents:
```

Comment:
Can you verify this is working as expected? The logic doesn't seem correct. Make sure there are tests for breaking inheritance with multiple rule types applying at the current resource level (self, children, self_and_children). It may be the change I suggested for _get_resource_rules broke the assumption of rule order implicitly made by this check.